src/Package/TestApp/Data/GridDataTest.cs

```diff
@@ -0,0 +1,257 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using System.Threading.Tasks;
+using FluentAssertions;
+using Microsoft.Common.Core.Test.Utility;
+using Microsoft.R.Host.Client;
+using Microsoft.R.Host.Client.Session;
+using Microsoft.R.Host.Client.Test.Script;
+using Microsoft.UnitTests.Core.XUnit;
+using Microsoft.UnitTests.Core.XUnit.MethodFixtures;
+using Microsoft.VisualStudio.R.Package.DataInspect;
+using Microsoft.VisualStudio.R.Package.DataInspect.DataSource;
+using Xunit;
+
+namespace Microsoft.VisualStudio.R.Interactive.Test.Data {
+    // TODO: determine and remove any duplicates from EvaluationWrapperTest.
+    // TODO: move this to Microsoft.R.Host.Client.Test after removing GridData dependencies on Package.
+    [ExcludeFromCodeCoverage]
+    public class GridDataTest : IAsyncLifetime {
+        private struct GridElement<T> {
+            public int X, Y;
+            public T Value;
+
+            public override string ToString() => $"[{Y}, {X}] = {Value}";
+        }
+
+        private readonly MethodInfo _testMethod;
+        private readonly IRSessionProvider _sessionProvider;
+        private readonly IRSession _session;
+
+        public GridDataTest(TestMethodFixture testMethod) {
+            _testMethod = testMethod.MethodInfo;
+            _sessionProvider = new RSessionProvider();
+            _session = _sessionProvider.GetOrCreate(Guid.NewGuid());
+        }
+
+        public async Task InitializeAsync() {
+            await _session.StartHostAsync(new RHostStartupInfo {
+                Name = _testMethod.Name,
+                RBasePath = RUtilities.FindExistingRBasePath()
+            }, new RHostClientTestApp(), 50000);
+        }
+
+        public async Task DisposeAsync() {
+            await _session.StopHostAsync();
+            _sessionProvider.Dispose();
+        }
+
+        private static IEnumerable<T> ToEnumerable<T>(IRange<T> range) {
+            int start = range.Range.Start, end = start + range.Range.Count;
+            for (int i = start; i < end; ++i) {
+                yield return range[i];
+            }
+        }
+
+        private void ShouldEqual<T>(IGrid<T> actual, T[,] expected) {
```

Comment:
I think we need a generic assertion for 2D arrays (including non-zero bounds). Then this code just needs a small helper to convert IGrid to one. I'll file that.